agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala

```diff
@@ -17,56 +17,87 @@
+ * @param atumPartitions
+ * @param parentAgent
+ * @param measures
  */
+class AtumContext private[agent](
+                                  val atumPartitions: AtumPartitions,
+                                  val parentAgent: AtumAgent,
+                                  private var measures: Set[Measure] = Set.empty) {
+
+  def currentMeasures: Set[Measure] = measures
+
+  def subPartitionContext(subPartitions: AtumPartitions): AtumContext = {
+    parentAgent.getOrCreateAtumSubContext(atumPartitions ++ subPartitions)(this)
+  }
 
-case class AtumContext(measurements: Set[Measurement] = Set()) {
+  def createCheckpoint(checkpointName: String, dataToMeasure: DataFrame) = {
+    ??? //TODO #26
+  }
 
-  def withMeasuresReplaced(
-    byMeasure: Measurement
-  ): AtumContext =
-    this.copy(measurements = Set(byMeasure))
+  def saveCheckpointMeasurements(checkpointName: String, measurements: Seq[Measure]) = {
+    ??? //TODO #55
+  }
 
-  def withMeasuresReplaced(
-    byMeasures: Iterable[Measurement]
-  ): AtumContext =
-    this.copy(measurements = byMeasures.toSet)
+  def addAdditionalData(key: String, value: String) = {
+    ??? //TODO #60
+  }
 
-  def withMeasuresAdded(
-    measure: Measurement
-  ): AtumContext =
-    this.copy(measurements = measurements + measure)
+  def addMeasure(newMeasure: Measure): AtumContext = {
+    measures = measures + newMeasure
+    this
+  }
 
-  def withMeasuresAdded(
-    measures: Iterable[Measurement]
-  ): AtumContext =
-    this.copy(measurements = measurements ++ measures)
+  def addMeasures(newMeasures: Set[Measure]): AtumContext = {
+    measures = measures ++ newMeasures
+    this
+  }
 
-  def withMeasureRemoved(measurement: Measurement): AtumContext =
-    this.copy(measurements = measurements.filterNot(_ == measurement))
+  def removeMeasure(measureToRemove: Measure): AtumContext = {
+    measures = measures - measureToRemove
+    this
+  }
 
+  private[agent] def copy(
+                           atumPartitions: AtumPartitions = this.atumPartitions,
+                           parentAgent: AtumAgent = this.parentAgent,
+                           measures: Set[Measure] = this.measures
+    ): AtumContext = {
+    new AtumContext(atumPartitions, parentAgent, measures)
+  }
 }
 
 object AtumContext {
+  type AtumPartitions = ListMap[String, String]
+
+  object AtumPartitions {
+    def apply(elems: (String, String)): AtumPartitions = {
+      ListMap(elems)
+    }
+  }
+
   implicit class DatasetWrapper(df: DataFrame) {
 
     /**
-     *  Executes the measure directly with not AtumContext.
+     *  Executes the measure directly (without AtumContext).
      *  @param measure the measure to be calculated
      *  @return
      */
-    def executeMeasure(checkpointName: String, measure: Measurement): DataFrame = {
-
+    def executeMeasure(checkpointName: String, measure: Measure): DataFrame = {
```

Comment:
I think that execute should do only executing and publishing should do publishing. And it can be all put together in some other operation. 

I definitely wouldn't expect from 'executeMeasure' to actually also send data to the server